Adding 3D barycentric coordinates for convex polyhedra

sloriot opened the pull request:
## Summary of Changes

New package for 3D barycentric coordinates in a convex polyhedron

## Release Management

* Affected package(s): New package
* Feature/Small Feature (if any): [here](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Barycentric_coordinates_3)
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership: GF

## TODO
- [x] check branch size (@sloriot)
- [x] check usage of `get_tolerance()` from `Barycentric_coordinates_3/include/CGAL/Barycentric_coordinates_3/internal/utils_3.h`

antoniospg committed:
adding basic file structure

antoniospg committed:
Created tetrahedron_coordinates_3 that contains the functions to calculate tetrahedra coordinates, as well as utils_3 that do the actual calculation, and one example to test the possible 3 cases of points

antoniospg committed:
Fixing cmakelist and updating package_info

antoniospg committed:
fixing cmakelists of test and benchmark

antoniospg committed:
Created structure for wachspress coordinates

antoniospg committed:
enhancing the code structure, ran some tests with circulators, property_maps and surface mesh

antoniospg committed:
better api, barycentric_enums and tests with tetrahedrons in the new api

danston committed:
making WP3 constructor work with default pmesh params

antoniospg committed:
first part of the calculation of wp coordinates

antoniospg committed:
FIRST FUNCTIONAL VERSIONgit add .git add .git add . Just implemented the equations, now I will make my code more organized

antoniospg committed:
Testing better examples and make code look cleaner

danston committed:
review 1 for wachspress coordinates 3

antoniospg committed:
fixing code issues after revision

antoniospg committed:
testing different kernels

antoniospg committed:
fixed minor issue

antoniospg committed:
added tetrahedron vs wp test and the beggining of the general one

antoniospg committed:
changing the name triangle_coordiantes_3 to tetrahedron_coordinates and removing -tol from the wp test

antoniospg committed:
first test with cube, testing unit partition, barycenter, weights between 0 and 1. Added utils.h to generate polyhedrons

antoniospg committed:
finishing test for cube, add generator of random points inside tetraherdron

antoniospg committed:
removed vscode file

antoniospg committed:
fixing tests for wachspress coordinates after review

antoniospg committed:
First implementation of DH coordinates (not working yet)

antoniospg committed:
fixing some bugs of the previous version

antoniospg committed:
functional version and first test of dh coordinates

antoniospg committed:
adding tests, fix for wp coordinates, dh now working for cube

antoniospg committed:
dh coordinates final adjustements

antoniospg committed:
draft version of mv coordinates

antoniospg committed:
fixing some bugs

antoniospg committed:
functional version of mv coordinates + tests

antoniospg committed:
First version voronoi coordinates (still debugging)

antoniospg committed:
Fixing after first review, need to finish tests

antoniospg committed:
more fixes and created first benchmark

antoniospg committed:
modifying api to handle edge cases

danston committed:
make it work with arbitrary face graphs + fixed warnings

antoniospg committed:
adding test to verify boundary

antoniospg committed:
first functional version of edge cases(only wp now)

antoniospg committed:
final version edge cases

antoniospg committed:
minor fix

antoniospg committed:
fixes after review + adding doxygen files

antoniospg committed:
final version edge_cases

antoniospg committed:
added shape_deformation example

antoniospg committed:
one more example and update docs

antoniospg committed:
adding final example

antoniospg committed:
fixed some examples + updated the docs

antoniospg committed:
first version docs

antoniospg committed:
fixing warnings in docs build

antoniospg committed:
fixing issues after first revision

antoniospg committed:
more fixes

antoniospg committed:
minor fixes

antoniospg committed:
adding thumb

antoniospg committed:
new line at the end of each file

antoniospg committed:
WP boundary coordinates can now be any polygon

antoniospg committed:
updated doc after first revision

antoniospg committed:
final tweaks

antoniospg committed:
Final fixes in the docs after revision

antoniospg committed:
adding benchmark scripts

sloriot committed:
use new macro

sloriot committed:
fix permission

sloriot committed:
update cmake scripts

sloriot committed:
add in package list

sloriot commented:
/build:v1

sloriot committed:
add license include header

sloriot committed:
remove tabs

sloriot committed:
avoid duplicated targets

sloriot committed:
fix compilation errors and do not use Surface_mesher

sloriot committed:
simplify

sloriot commented:
cc @antoniospg :)

danston commented:
@sloriot No, for center query in a cube there are no edge cases. The sum should be strictly positive because all weights are positive. My guess would be that certain weights are positive and certain are negative so they sum up to zero. If this is a case, most likely the orientation of faces is different from what it has been when the test was developed because it did pass when Antonio worked on it. Could that be that this line
```
CGAL::make_hexahedron(p0, p1, p2, p3, p4, p5, p6, p7, polyhedron, CGAL::parameters::do_not_triangulate_faces(false));
```
produces different results from this one:
```
PMP::triangulate_faces(faces(polyhedron), polyhedron);
```
In any case for such a simple test, I would first check how the cube looks like, orientation of its faces, and values of the weights.

danston commented on Barycentric_coordinates_3/include/CGAL/Barycentric_coordinates_3/Wachspress_coordinates_3.h:
```diff
@@ -0,0 +1,398 @@
+
+    // Compute wp coordinates for a given vertex v and a query q
+    template<typename Vertex>
+    FT compute_wp_vertex_query(const Vertex& vertex, const Point_3& query){
```
@sloriot The issue is probably coming from this method. There is suspicious negation of the first face normal, which is not coherent with other normals used here. Implementations of other MV and DH coordinates do not have this. My guess would be that this method does not return the same weights for the same mesh but differently oriented faces, which should not be the case. What is important here is to have consistent normals as indicated in the [original paper](https://www.cs.wustl.edu/~taoju/research/generalbary_new.pdf). Maybe just try to remove the negation in line 269 and see what happens? 

There are multiple formulas to compute WP coordinates. I do not remember which one Antonio used here. Maybe it is worth using all formulas from the same paper to be consistent. Finally, another test that is easy to do is to change WP to DH coordinates in this test and see if it works. If I remember correctly for such a simple case (center point in a cube) WP and DH should produce exactly the same result.

sloriot replied:
I went back to Meyers' paper and updated the formula. This test passes but now  `test_dh_weights_3` and `test_edge_cases` don't.  TBC

sloriot replied:
Fixed!

sloriot committed:
add missing include

sloriot committed:
fix weight computation

sloriot committed:
do not use deprecated functions

sloriot committed:
remove global normalization added in 4b7c3576a48

and add it locally. It is responsible for precision issues
for some assertions and it would also need a check everywhere
it is used to make sure it is intended

sloriot committed:
weight is negative is query is outside of the mesh

sloriot committed:
missing include

sloriot committed:
property_map_selector is not documented

sloriot committed:
cosmetics

sloriot committed:
fix dependencies

sloriot committed:
Merge branch 'master' into gsoc2021-Barycentric_coordinates_3-antoniospg

soesau commented:
/build:v0

soesau committed:
fixing example names

soesau committed:
fixing compilation issues

soesau commented:
/build:v0

soesau committed:
added note that calculated coordinates are not normalized

soesau commented:
/build:v0

soesau committed:
update example

soesau committed:
Merge branch 'main' into gsoc2021-Barycentric_coordinates_3-antoniospg

soesau commented:
/build:v0

soesau committed:
integrating review

soesau committed:
more renaming

soesau committed:
integrating review

soesau committed:
integrating review

soesau committed:
integrating review

soesau committed:
integrated last points from Efis review

soesau commented:
/build:v0

soesau committed:
fix 2 warnings

soesau committed:
integrating review

soesau committed:
integrated review

soesau committed:
bugfix

soesau committed:
bugfix

afabri commented on Barycentric_coordinates_3/examples/Barycentric_coordinates_3/tetrahedron_coordinates.cpp:
```diff
@@ -0,0 +1,41 @@
+#include <CGAL/Simple_cartesian.h>
+#include <CGAL/Barycentric_coordinates_3/tetrahedron_coordinates.h>
+
+//Typedefs
```
```suggestion
```

afabri commented on Barycentric_coordinates_3/examples/Barycentric_coordinates_3/shape_deformation_3.cpp:
```diff
@@ -0,0 +1,71 @@
+using FT =  Kernel::FT;
+
+using Surface_mesh =  CGAL::Surface_mesh<Point_3>;
+namespace PMP = CGAL::Polygon_mesh_processing;
```
```suggestion
```

soesau committed:
doc fix

soesau committed:
clean up examples

soesau commented:
/build:v0

soesau committed:
entry in CHANGES.md
[skip ci]

soesau committed:
Merge branch 'main' into gsoc2021-Barycentric_coordinates_3-antoniospg

sloriot committed:
fix compilation errors and warnings

sloriot committed:
use tolerance

soesau committed:
updating package version

soesau committed:
introducing named parameters
adding test for concept
cleaning up examples

soesau committed:
pass on concepts

soesau committed:
adding Equal_2 to Projection_traits_base_3

soesau committed:
adding more functors to concept

soesau committed:
pass on concept

soesau committed:
removing tabs
reordering template parameters in 3D barycentric coordinates classes

soesau committed:
fixing reference pages
integrating reviews

sloriot committed:
fix warnings and compilation errors

sloriot committed:
fix regression in test_edge_cases.cpp

restore the tolerance changed in 1432f40

soesau committed:
fixed test (removed generated points outside of tetrahedron)

soesau committed:
unique test name

sloriot committed:
move file
